contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+-   **Example Usage**: amp-date-countdown, amp-date-display
+
+#### init
+
+-   **Default**: Optional.
+-   **Override**: Sometimes, if additional processing is needed in the DOM.
+-   **Usage**: If your extension 'props' definition is not sufficient, or when registering AMP actions and events.
+-   **Example Usage**: amp-base-carousel, amp-lightbox
+
+#### mutationObserverCallback
+
+-   **Default**: Optional.
+-   **Override**: Sometimes, if additional processing is needed to make the DOM mutable.
+-   **Usage**: A callback called immediately after mutations have been observed on a component. This differs from `checkPropsPostMutations` in that it is called in all cases of mutation, not just the subset of attributes configured in the extension 'props' definition.
+-   **Example Usage**: amp-base-carousel, amp-sidebar
+
+#### updatePropsForRendering
+
+-   **Default**: Optional.
+-   **Override**: Sometimes, if additional processing is needed in the DOM.
+-   **Usage**: A callback called to compute props before rendering is run. The properties computed here are ephemeral and thus should not be persisted via a `mutateProps()` method.
+-   **Example Usage**: amp-timeago
+
+### AMP/Preact Bridge
+
+#### PreactBaseElement['Component']
+
+-   **Default**: Required.
+-   **Override**: Always.
+-   **Usage**: Define the corresponding Preact functional component.
+-   **Example Usage**: amp-fit-text, amp-timeago
+
+#### PreactBaseElement['props']
+
+-   **Default**: Optional.
+-   **Override**: Almost always.
+-   **Usage**: Define the mapping of Preact prop to AmpElement DOM attributes. These will update and re-render the component on DOM mutation.
+-   **Example Usage**: amp-base-carousel, amp-lightbox
+
+#### PreactBaseElement['staticProps']
+
+-   **Default**: Optional.
+-   **Override**: Almost always.
+-   **Usage**: Define the mapping of Preact prop to AmpElement DOM attributes. These will not update or re-render the component on DOM mutation. Can be used in lieu of init().
+-   **Example Usage**: n/a
+
+#### PreactBaseElement['className']
+
+-   **Default**: Optional.
+-   **Override**: Rarely.
+-   **Usage**: Specify an exact className prop to Preact.
+-   **Example Usage**: n/a
+
+#### PreactBaseElement['layoutSizeDefined']
+
+-   **Default**: Optional.
+-   **Override**: Sometimes.
+-   **Usage**: Specify that the component requires `layoutSizeDefined`, meaning it can have a layout of any of the following: `fixed`, `fixed-height`, `responsive`, `fill`, `flex-item`, `fluid`, or `intrinsic`. Learn more about how to use this in [Layouts supported in your element](#layouts-supported-in-your-element) or generally with [layouts in AMP](https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout).
+-   **Example Usage**: n/a
+
+#### PreactBaseElement['loadable']
+
+-   **Default**: Optional.
+-   **Override**: Sometimes, when the element implements a loading protocol.
```

Comment:
IMHO we should say "Required for any component that makes network requests (directly or via other DOM elements). Optional otherwise."